app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts

```diff
@@ -17,6 +20,96 @@ const appendSwaggerDiv = (container: HTMLElement, id: string): void => {
   container.appendChild(div)
 }
 
+/**
+ * when using Record notation, the following error is thrown:
+ * 'TS2456: Type alias 'BodyValue' circularly references itself.'
+ */
+// eslint-disable-next-line @typescript-eslint/consistent-indexed-object-style
+type BodyValue = boolean | number | string | { [key: string]: BodyValue }
+
+/**
+ * Transforms an object into form data representation, also URL-encoding the values,
+ * Example:
+ * {
+ *   a_string: 'hello',
+ *   an_array: [
+ *     { first: 1 },
+ *     { second: 1, extra_param: 'with whitespace'}
+ *   ]
+ * }
+ * =>
+ * {
+ *   a_string: 'hello',
+ *   'an_array[0][first]': '1',
+ *   'an_array[1][second]': '1',
+ *   'an_array[1][extra_param]': 'with%20whitespace'
+ * }
+ * @param object
+ */
+export const objectToFormData = (object: BodyValue): Record<string, boolean | number | string> => {
+  const buildFormData = (formData: Record<string, boolean | number | string>, data: BodyValue, parentKey?: string) => {
+    if (data && typeof data === 'object') {
+      Object.keys(data).forEach((key: string) => {
+        buildFormData(formData, data[key], parentKey ? `${parentKey}[${key}]` : key)
+      })
+    } else {
+      if (parentKey) {
+        formData[parentKey] = data ? data : ''
+      }
+    }
+  }
+  const formData = {}
+  buildFormData(formData, object)
+  return formData
+}
```

Comment:
I spent some time trying to understand this and I still don't understand:

1. How does this transforms an array into the `'an_array[0][first]': '1'` form? Are arrays being treated as objects?
2. How this URL encodes? where?

I think I need some comments here.

Reply:
> 1. How does this transforms an array into the 'an_array[0][first]': '1' form? Are arrays being treated as objects?

Yes, the trick is in how JS treats arrays, basically it's like an object where the keys are the array indexes. So, Object.keys(myArray) returns ['0', '1', '2'] etc. See:
![Screenshot from 2023-04-11 11-18-14](https://user-images.githubusercontent.com/1270328/231115409-0d08ef40-ee90-4014-9f0a-523e07844e9f.png)

> 2. How this URL encodes? where?

I actually had some back-and-forth with this one, I tried applying `encodeURIComponent()` to the value (`formData[parentKey] = data ? encodeURIComponent(data) : ''`), but then I realised that in this case there was double-encoding (e.g. spaces looked like `%20%2520` or smth like that). I think, the original `execute` method of the `swagger-client` that we're calling from the plugin takes care of the encoding (I guess [these methods](https://github.com/swagger-api/swagger-js/blob/v3.15.0/src/http/index.js#L363-L405) are used).